api/insight/socket.io.go

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2017, The dcrdata developers
+// See LICENSE for details.
+
+package insight
+
+import (
+	"github.com/googollee/go-socket.io"
+
+	"github.com/decred/dcrd/dcrjson"
+	"github.com/decred/dcrd/wire"
+	"github.com/decred/dcrdata/blockdata"
+	"github.com/decred/dcrdata/txhelpers"
+)
+
+type SocketServer struct {
+	socketio.Server
+}
+
+func NewSocketServer(newTxChan chan *NewTx) (*SocketServer, error) {
+	server, err := socketio.NewServer(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	//server.BroadcastTo()
+	server.On("connection", func(so socketio.Socket) {
+		apiLog.Debug("New socket.io connection")
+
+		so.On("disconnection", func() {
+			apiLog.Debug("Client disconnected")
+		})
+		so.On("subscribe", func(room string) {
+			apiLog.Debugf("Client joining room: %s", room)
+			so.Join(room)
+		})
+	})
+
+	server.On("error", func(so socketio.Socket, err error) {
+		apiLog.Errorf("Insight socket server error: %v", err)
+	})
+
+	sockServ := SocketServer{*server}
+	go sockServ.sendNewTx(newTxChan)
+	return &sockServ, nil
+}
+
+func (soc *SocketServer) Store(blockData *blockdata.BlockData, _ *wire.MsgBlock) error {
+	block := WebSocketBlock{blockData.Header.Hash}
+	soc.BroadcastTo("inv", "block", block)
+	return nil
+}
+
+func (soc *SocketServer) sendNewTx(newTxChan chan *NewTx) {
+	for {
+		ntx := <-newTxChan
+		msgTx, _ := txhelpers.MsgTxFromHex(ntx.Hex)
+		vouts := make(map[string]int64)
+		var total int64
+		isStake := txhelpers.IsStakeTx(msgTx)
+		for i, v := range msgTx.TxOut {
+			total += v.Value
+			if !isStake && len(ntx.Vouts[i].ScriptPubKey.Addresses) != 0 {
+				vouts[ntx.Vouts[i].ScriptPubKey.Addresses[0]] = v.Value
+			}
+		}
+		tx := WebSocketTx{
+			Hash:     msgTx.TxHash().String(),
+			Size:     len(ntx.Hex) / 2,
+			TotalOut: total,
+			Vout:     vouts,
+		}
+		soc.BroadcastTo("inv", "tx", tx)
```

Comment:
I like this concept of "rooms" that socket.io brings.  Much easier than our hand-rolled websocket solution.

Please be sure to try working with the go-socket.io v1.4 branch as recommended by their README.md as apparently the project is abandoned by the maintainer.